src/model.cpp

```diff
@@ -130,23 +130,42 @@ Model::Model(const int nx_rdata, const int nxtrue_rdata, const int nx_solver,
        if Python wrapped: dxdotdp_explicit and dxdotdp_implicit are CSC matrices
      */
     if (pythonGenerated) {
-        dxdotdp_explicit =
-            SUNMatrixWrapper(nx_solver, static_cast<int>(p.size()),
-                             ndxdotdp_explicit, CSC_MAT);
+
+        dwdw_ = SUNMatrixWrapper(nw, nw, ndwdw, CSC_MAT);
+        // size dynamically adapted for dwdx_ and dwdp_
+        dwdx_ = SUNMatrixWrapper(nw, nx_solver, 0, CSC_MAT);
+        dwdp_ = SUNMatrixWrapper(nw, static_cast<int>(p.size()), 0, CSC_MAT);
+
+        for (int irec = 0; irec <= w_recursion_depth_; ++irec) {
+            dwdp_hierarchical_.emplace_back(
+                SUNMatrixWrapper(nw, static_cast<int>(p.size()), ndwdw + ndwdp,
+                                 CSC_MAT)); // guessed size
+            dwdx_hierarchical_.emplace_back(SUNMatrixWrapper(
```

Comment:
If I get it correctly, this is computationally more heavy than the current implementation, but will be more efficient for large models during simulation, right?
Can we quantify this somehow?

Reply:
this only applies if `w` is recursive, otherwise (which is the case for sbml models with conservation laws) there should be 0 overhead.

Reply:
> this only applies if `w` is recursive, otherwise (which is the case for sbml models with conservation laws) there should be 0 overhead.

i.e. if this turns out to be a problem, an easy fix would be to just flatten `w` in ode_export, in which case w_recursion_depth_ would be 0.